src/bigint.rs

```diff
@@ -467,146 +522,303 @@ impl Unsigned for BigUint {}
+        } else if borrow != 0 {
+            *ai = sbb(*ai, 0, &mut borrow);
+        } else {
+            break;
         }
+    }
 
-        for a in &mut diff[other.len()..] {
-            if borrow == 0 { break }
-            let d = big_digit::BASE - borrow
-                + (*a as DoubleBigDigit);
-            let (hi, lo) = big_digit::from_doublebigdigit(d);
-            borrow = if hi == 0 { 1 } else { 0 };
-            *a = lo;
-        }
+    /* note: we're _required_ to fail on underflow */
+    assert!(borrow == 0 && b_iter.all(|x| *x == 0),
+            "Cannot subtract b from a because b is larger than a.");
+}
+
+impl<'a> Sub<&'a [BigDigit]> for BigUint {
+    type Output = BigUint;
+
+    fn sub(mut self, other: &[BigDigit]) -> BigUint {
+        sub2(&mut self.data[..], other);
+        self.normalize()
+    }
+}
+
+impl<'a> Sub<&'a BigUint> for BigUint {
+    type Output = BigUint;
 
-        assert!(borrow == 0, "arithmetic operation overflowed");
-        BigUint::new(diff)
+    fn sub(self, other: &BigUint) -> BigUint {
+        self - &other.data[..]
     }
 }
 
+fn sub_sign(a: &[BigDigit], b: &[BigDigit]) -> BigInt {
+    fn cmp(a: &[BigDigit], b: &[BigDigit]) -> Ordering {
+        /*
+         * Since we're working with slices here, it's _not_ guaranteed that the highest element in
+         * the slice is nonzero:
+         */
+        let mut i = cmp::max(a.len(), b.len());
 
-forward_all_binop_to_val_ref_commutative!(impl Mul for BigUint, mul);
+        while {
+            let av = if i < a.len() { a[i] } else { 0 };
+            let bv = if i < b.len() { b[i] } else { 0 };
 
-impl<'a> Mul<&'a BigUint> for BigUint {
-    type Output = BigUint;
+            if av < bv { return Less; }
+            if av > bv { return Greater; }
 
-    fn mul(self, other: &BigUint) -> BigUint {
-        if self.is_zero() || other.is_zero() { return Zero::zero(); }
-
-        let (s_len, o_len) = (self.data.len(), other.data.len());
-        if s_len == 1 { return mul_digit(other.clone(), self.data[0]);  }
-        if o_len == 1 { return mul_digit(self,  other.data[0]); }
-
-        // Using Karatsuba multiplication
-        // (a1 * base + a0) * (b1 * base + b0)
-        // = a1*b1 * base^2 +
-        //   (a1*b1 + a0*b0 - (a1-b0)*(b1-a0)) * base +
-        //   a0*b0
-        let half_len = cmp::max(s_len, o_len) / 2;
-        let (s_hi, s_lo) = cut_at(self,  half_len);
-        let (o_hi, o_lo) = cut_at(other.clone(), half_len);
-
-        let ll = &s_lo * &o_lo;
-        let hh = &s_hi * &o_hi;
-        let mm = {
-            let (s1, n1) = sub_sign(s_hi, s_lo);
-            let (s2, n2) = sub_sign(o_hi, o_lo);
-            match (s1, s2) {
-                (Equal, _) | (_, Equal) => &hh + &ll,
-                (Less, Greater) | (Greater, Less) => &hh + &ll + (n1 * n2),
-                (Less, Less) | (Greater, Greater) => &hh + &ll - (n1 * n2)
-            }
-        };
+            i != 0
+        } { i = i - 1; }
```

Comment:
On Mon, Nov 30, 2015 at 03:40:51PM -0800, Josh Stone wrote:
> > -        let (s_hi, s_lo) = cut_at(self,  half_len);
> > -        let (o_hi, o_lo) = cut_at(other.clone(), half_len);
> > -
> > -        let ll = &s_lo * &o_lo;
> > -        let hh = &s_hi * &o_hi;
> > -        let mm = {
> > -            let (s1, n1) = sub_sign(s_hi, s_lo);
> > -            let (s2, n2) = sub_sign(o_hi, o_lo);
> > -            match (s1, s2) {
> > -                (Equal, _) | (_, Equal) => &hh + &ll,
> > -                (Less, Greater) | (Greater, Less) => &hh + &ll + (n1 * n2),
> > -                (Less, Less) | (Greater, Greater) => &hh + &ll - (n1 * n2)
> > -            }
> > -        };
> > +            i != 0
> > +        } { i = i - 1; }
> 
> Weird, I didn't even know you could write a while-loop this way...
> 
> Anyway, it seems wasteful not to use `cmp_slice`.  A quick experiment to just normalize the zeros away first seems to work just as fast:
> 
> ```
>     let a_len = a.iter().rposition(|&x| x != 0).map_or(0, |i| i + 1);
>     let b_len = b.iter().rposition(|&x| x != 0).map_or(0, |i| i + 1);
>     match cmp_slice(&a[..a_len], &b[..b_len]) { ... }
> ```

Oh, I like that.